### PR TITLE
JAMES-3756 JMAP Delegation extension specification

### PR DESCRIPTION
**Pull request by quantranhong1999**

### CODE REVIEW

File: server/protocols/jmap-rfc-8621/doc/specs/spec/mail/delegation.mdown [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+A **Delegation** object stores information about an email address that either has access to the user's account or delegates access
+to the user. The semantics for this object depends on the context of usage methods: either manage delegate accounts or delegated accounts.
+
+The delegation object MUST have the following properties:
+- **id**: `Id` (immutable; server-set)
+  The id of the Delegation.
+- **username**: `String` (immutable)
+  The email address of the user that either delegates or is delegated to this user account.
+
+A user can have multiple delegate accounts as well as multiple delegated accounts.
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/doc/specs/spec/mail/delegation.mdown:36]: I find this sentence confusing... What about:

`A user can delegate his account to multiple users as well as have access to the account of multiple users.` 

Or something like that?

File: server/protocols/jmap-rfc-8621/doc/specs/spec/mail/delegation.mdown [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+A user can create, delete and retrieve his delegate accounts while only can delete and retrieve his delegated accounts.
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/doc/specs/spec/mail/delegation.mdown:38]: Find this a bit confusing too like above

File: server/protocols/jmap-rfc-8621/doc/specs/spec/mail/delegation.mdown [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+A server MUST not allow a delegated user to access and manage the owner user's delegation setting to avoid privilege escalation.
+
+The following JMAP methods are supported:
+
+### Delegate/get
+This method returns accounts that the user delegates to access his account.
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/doc/specs/spec/mail/delegation.mdown:45]:
```suggestion
This method returns accounts that the user has delegated access to.
```
?

File: server/protocols/jmap-rfc-8621/doc/specs/spec/mail/delegation.mdown [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+This is a standard "/get" method as described in [@!RFC8620], Section 5.1. The *ids* argument may be `null` to fetch all at once.
+
+A server MUST return the *SetError* type `forbidden` when a delegated user accesses the owner user's delegation setting.
+
+### Delegate/set
+This method allows the user to manage (create and delete) accounts that he delegates to access his account.
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/doc/specs/spec/mail/delegation.mdown:52]:
```suggestion
This method allows the user to manage (create and delete) accounts that he delegates its access to.
```

File: server/protocols/jmap-rfc-8621/doc/specs/spec/mail/delegation.mdown [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+This is a standard "/set" method as described in [@!RFC8620], Section 5.3, except the update part because of its limited usage.
+
+A server MUST return the *SetError* type `forbidden` when a delegated user accesses the owner user's delegation setting.
+
+### DelegatedAccount/get
+This method returns accounts that the user is delegated to access.
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/doc/specs/spec/mail/delegation.mdown:59]:
```suggestion
This method returns accounts that the user has access to.
```

File: server/protocols/jmap-rfc-8621/doc/specs/spec/mail/delegation.mdown [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+
+This is a standard "/get" method as described in [@!RFC8620], Section 5.1. The *ids* argument may be `null` to fetch all at once.
+
+A server MUST return the *SetError* type `forbidden` when a delegated user accesses the owner user's delegation setting.
+
+### DelegatedAccount/set
+This method allows the user to delete accounts that he is delegated to access.
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/doc/specs/spec/mail/delegation.mdown:66]:
```suggestion
This method allows the user to delete accounts that he has access to.
```

File: server/protocols/jmap-rfc-8621/doc/specs/spec/mail/intro.mdown [hunk 7]
```diff
@@ -153,6 +153,17 @@ This extension is specific to the Apache James server and enables to:
...
+### urn:apache:james:params:jmap:delegation
+
+> :information_source:
+> Not implemented
+
+This represents support for the Delegation data type and associated API methods. Servers supporting this specification MUST add this property to the capabilities object.
+
+This extension is specific to the Apache James server and enables to:
+
+* manage accounts that a user delegates to access his account, and manage accounts that a user is delegated to access.
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/doc/specs/spec/mail/intro.mdown:165]:
```suggestion
* manage accounts that a user delegated access to, and accounts that a user is allowed to access.
```

**Comment by vttranlina** [comment]
read it

**Review by Arsnael — APPROVED** [review]

File: server/protocols/jmap-rfc-8621/doc/specs/spec/mail/delegation.mdown [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+The delegation object MUST have the following properties:
+- **id**: `Id` (immutable; server-set)
+  The id of the Delegation.
+- **username**: `String` (immutable)
+  The email address of the user that either delegates or is delegated to this user account.
+
+A user can delegate his account to multiple users as well as be allowed to access other accounts.
+
+A user can manage his delegations to other users to access his account by creating, deleting and retrieving delegation objects.
+Meanwhile, a user can only retrieve and delete the accesses delegated to him.
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/doc/specs/spec/mail/delegation.mdown:39]: IMO we miss a clear description of the two objects stated below.

```
The **Delegate** object follows the structure of **Delegation** and represents delegations accessing this account (allowing the listed user to connect as the current account owner).

The **DelegatedAccount** object object follows the structure of **Delegation** and represents delegations active for this account (allowing the current account owner to connect as the listed user).
```

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Pull request merged**